Occlusion sensitivity

rijobro opened the pull request:
Calculate the occlusion sensitivity of an image. Block part of the image, and record how the probability of the inferred class changes. Then iteratively move the occluded section across the image to build up a sensitivity map.

Once merged, I will add a small example to the end of the [2D classification tutorial](https://github.com/Project-MONAI/tutorials/blob/master/2d_classification/mednist_tutorial.ipynb). 

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

wyli commented on monai/metrics/occlusion_sensitivity.py:
```diff
@@ -0,0 +1,116 @@
+    be useful to understand why a network is making certain decisions.
+
+    Args:
+        model: model to use for inference
```
could you specify the expected shape for the model input and output and the labels? if it supports 3D, would be great to have a test case

wyli replied:
or maybe it's GitHub's issue https://www.githubstatus.com/incidents/k46g5dn2c4lt ...

wyli reviewed (commented):
thanks!  very interesting metrics, perhaps could include a paper reference in the docstring about the method?

casperdcl commented on monai/metrics/occlusion_sensitivity.py:
```diff
@@ -0,0 +1,120 @@
+
+        # Clone and replace target area with `pad_val`
+        occlu_im = image.clone()
+        occlu_im[..., min_idx[0] : max_idx[0], min_idx[1] : max_idx[1], min_idx[2] : max_idx[2]] = pad_val
```
For arbitrary `ndim`

```suggestion
        occlu_im[(...,) + tuple(slice(i, j) for i, j in zip(min_idx, max_idx))] = pad_val
```

rijobro commented:
@wyli Thanks for the review. I've updated the documentation with expected input and outputs and added a 3D example. I've also examples in a [PR](https://github.com/Project-MONAI/tutorials/pull/47) in the Tutorials repository.

Once the tests all pass, I'm happy for this to be merged. Do you think you could give it another look over?

wyli commented on monai/metrics/occlusion_sensitivity.py:
```diff
@@ -0,0 +1,188 @@
+    num_required_predictions = np.prod(output_im_shape)
+
+    # Loop 1D over image
+    for i in trange(num_required_predictions):
```
looks like there's some memory issue with the Windows test, I don't know the root cause for now... is it possible that this for loop is actually accumulating the model variable gradients, using a lot of memory? adding some `detach()` commands here might help?

rijobro replied:
The Windows test was originally failing but I decreased batch size and that seems to be passing now [link](https://github.com/Project-MONAI/MONAI/pull/1099/checks?check_run_id=1283108651#step:9:2582). The current error on Windows looks unrelated. I'll relaunch and see what happens.

wyli reviewed (approved):
thanks!
